File: src/libs/jderobotcomm_cpp/src/ice/cameraIceClient.cpp
/*
 *  Copyright (C) 1997-2016 JDE Developers TeamkinectViewer.camRGB
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see http://www.gnu.org/licenses/.
 *
 *  Author : Jose María Cañas <jmplaza@gsyc.es>
			Francisco Miguel Rivas Montero <franciscomiguel.rivas@urjc.es>

 */

#include "jderobot/comm/ice/cameraIceClient.hpp"

#include <exception>


namespace JdeRobotComm {


CameraIceClient::CameraIceClient(JdeRobotComm::Communicator* jdrc, std::string prefix) {

	this->prefix=prefix;
	Ice::ObjectPrx baseCamera;
	this->refreshRate=0;
	this->mImageFormat.empty();

	

	float fps=jdrc->getConfig().asFloatWithDefault(prefix+".Fps", 30);
	this->cycle=(1/fps)*1000000;
	try{
		std::string proxy = jdrc->getConfig().asString(prefix+".Proxy");
		baseCamera = jdrc->getIceComm()->stringToProxy(proxy);
		if (0==baseCamera){
			this->on = false;
			throw prefix + "Could not create proxy with Camera";
		}
		else {
			this->prx= jderobot::CameraPrx::checkedCast(baseCamera);
			this->on = true;
			if (0==this->prx){
				this->on = false;
				throw "Invalid " + prefix + ".Proxy";
			}
		}
	}catch (const Ice::Exception& ex) {
		std::cerr << ex << std::endl;
	}
	catch (const char* msg) {
		std::cerr << msg << std::endl;
		LOG(ERROR) <<prefix + " Not camera provided";
	}

	//check if default format is defined
	std::string definedFormat=jdrc->getConfig().asStringWithDefault(prefix+".Format", "RGB8");

	this->mImageFormat = CameraUtils::negotiateDefaultFormat(this->prx,definedFormat);

	jderobot::ImageDataPtr data = this->prx->getImageData(this->mImageFormat);

	this->pauseStatus=false;
}


CameraIceClient::~CameraIceClient() {
	this->on=false;
}


jderobot::ImageFormat CameraIceClient::getImageFormat()
{
	return (this->prx->getImageFormat());
}

void CameraIceClient::setImageFormat (std::string format)
{
	mImageFormat = format;

	LOG(INFO) <<"Changed format " + this->mImageFormat + " for camera " + this->prx->getCameraDescription()->name;
};


void CameraIceClient::reset(){
	this->prx->reset();
}

void CameraIceClient::pause(){
	this->pauseStatus=true;
}

void CameraIceClient::resume(){
	this->controlMutex.lock();
	this->pauseStatus=false;
	this->semWait.broadcast();
	this->controlMutex.unlock();
}


void
CameraIceClient::run(){
	jderobot::ImageDataPtr dataPtr;
	colorspaces::Image::FormatPtr fmt;
	IceUtil::Time last;

	int iterIndex = 0;
	int totalRefreshRate = 0;
	int refrRate = 0;

	JdeRobotTypes::Image img;

	last=IceUtil::Time::now();
	while (this->on){

		iterIndex ++;
		if (pauseStatus){
			IceUtil::Mutex::Lock sync(this->controlMutex);
			this->semWait.wait(sync);
		}

		try{

			
			dataPtr = this->prx->getImageData(this->mImageFormat);

			

			

			// Putting image data
			img.data = CameraUtils::getImageFromCameraProxy(dataPtr);
			
			img.format = dataPtr->description->format;
			img.width = dataPtr->description->width;
			img.height = dataPtr->description->height;
			img.timeStamp = dataPtr->timeStamp.seconds + dataPtr->timeStamp.useconds * 1e-6;




		}
		catch(std::exception& e){
			LOG(WARNING) << prefix +"error during request (connection error): " << e.what() << std::endl;
			usleep(50000);

		}

		int process = (IceUtil::Time::now().toMicroSeconds() - last.toMicroSeconds());



		if (process > (int)cycle ){
			DLOG(WARNING) << "--------" + prefix + " adquisition timeout-";
		}
		else{
			int delay = (int)cycle - process;
			if (delay <1 || delay > (int)cycle)
				delay = 1;

			usleep(delay);
		}


		int rate =(int)(1000000/(IceUtil::Time::now().toMicroSeconds() - last.toMicroSeconds()));
		totalRefreshRate =  totalRefreshRate + rate;
		refrRate = totalRefreshRate / iterIndex;
		last=IceUtil::Time::now();

		if (iterIndex == INT_MAX) 
		{
			iterIndex = 0;
			DLOG(INFO) << "*** Counter reset";
		}

		this->controlMutex.lock();
		this->image = img;
		this->refreshRate = refrRate;
		this->controlMutex.unlock();

	}

	this->image.data.release();
}

void CameraIceClient::stop_thread()
{
	this->on=false;
}

JdeRobotTypes::Image CameraIceClient::getImage(){
	JdeRobotTypes::Image img;

	this->controlMutex.lock();
	img = this->image;
	this->controlMutex.unlock();

	return img;
}

int CameraIceClient::getRefreshRate(){
	int rr;
	this->controlMutex.lock();
	rr = this->refreshRate;
	this->controlMutex.unlock();

	return rr;
};

} /* namespace jderobot */

File: src/libs/comm_py/comm/ice/cameraIceClient.py
#
#  Copyright (C) 1997-2017 JDE Developers Team
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see http://www.gnu.org/licenses/.
#  Authors :
#       Aitor Martinez Fernandez <aitor.martinez.fernandez@gmail.com>
#

import traceback
import jderobot
import numpy as np
import threading
import Ice
from .threadSensor import ThreadSensor
from jderobotTypes import Image


class Camera:
    '''
        Camera Connector. Recives image from Ice interface when you run update method.
    '''

    def __init__(self, jdrc, prefix):
        '''
        Camera Contructor.
        Exits When it receives a Exception diferent to Ice.ConnectionRefusedException

        @param jdrc: Comm Communicator
        @param prefix: Name of client in config file

        @type jdrc: Comm Communicator
        @type prefix: String
        '''

        self.lock = threading.Lock()
        self.image = Image()

        try:

            ic = jdrc.getIc()
            proxyStr = jdrc.getConfig().getProperty(prefix+".Proxy")
            basecamera = ic.stringToProxy(proxyStr)
            self.proxy = jderobot.CameraPrx.checkedCast(basecamera)
            self.imgFormat = jdrc.getConfig().getPropertyWithDefault(prefix+".Format", "RGB8")

            self.update()

            if not self.proxy:
                print ('Interface ' + prefix + ' not configured')

        except Ice.ConnectionRefusedException:
            print(prefix + ': connection refused')

        except:
            traceback.print_exc()
            exit(-1)

    def update(self):
        '''
        Updates Image.
        '''
        if self.hasproxy():
            img = Image()
            imageData = self.proxy.getImageData(self.imgFormat)
            img.height = imageData.description.height
            img.width = imageData.description.width
            img.format = imageData.description.format

            img.data = np.frombuffer(imageData.pixelData, dtype=np.uint8)
            img.data.shape = img.height, img.width, 3

            img.timeStamp = imageData.timeStamp.seconds + imageData.timeStamp.useconds * 1e-9


            self.lock.acquire()
            self.image = img
            self.lock.release()

    def getImage(self):
        '''
        Returns last Image. 

        @return last JdeRobotTypes Image saved

        '''
        img = Image()      
        if self.hasproxy():
            self.lock.acquire()
            img = self.image
            self.lock.release()
        return img

    def hasproxy (self):
        '''
        Returns if proxy has ben created or not. 

        @return if proxy has ben created or not (Boolean)

        '''
        return hasattr(self,"proxy") and self.proxy




class CameraIceClient:
    '''
        Camera Ice Client. Recives image from Ice interface running camera update method in a thread.
    '''
    def __init__(self,jdrc,prefix, start = False):
        '''
        CameraIceClient Contructor.

        @param jdrc: Comm Communicator
        @param prefix: Name of client in config file
        @param start: indicates if start automatically the client

        @type jdrc: Comm Communicator
        @type prefix: String
        @type start: Boolean
        '''
        self.camera = Camera(jdrc,prefix)

        self.kill_event = threading.Event()
        self.thread = ThreadSensor(self.camera, self.kill_event)
        #self.thread = ThreadSensor(self.camera)
        self.thread.daemon = True

        if start:
            self.start()

    
    def start(self):
        '''
        Starts the client. If client is stopped you can not start again, Threading.Thread raised error

        '''
        self.kill_event.clear()
        self.thread.start()

    
    def stop(self):
        '''
        Stops the client. If client is stopped you can not start again, Threading.Thread raised error

        '''
        self.kill_event.set()

    def getImage(self):
        '''
        Returns last Image. 

        @return last JdeRobotTypes Image saved

        '''
        return self.camera.getImage()


    def hasproxy (self):
        '''
        Returns if proxy has ben created or not. 

        @return if proxy has ben created or not (Boolean)

        '''
        return self.camera.hasproxy()
